switchapi/src/switch_packet.c

```diff
@@ -156,41 +183,198 @@ switch_packet_rx_from_hw()
+    switch_packet_tx_entry_t           tx_entry;
+    switch_packet_tx_info_t           *tx_info = NULL;
+    uint16_t                           ether_type = 0;
+    uint16_t                           offset = 0;
+    uint16_t                           vlan_offset = 0;
+    switch_status_t                    status = SWITCH_STATUS_SUCCESS;
+
+    memset(in_packet, 0x0, SWITCH_PACKET_MAX_BUFFER_SIZE);
+    memset(packet, 0x0, SWITCH_PACKET_MAX_BUFFER_SIZE);
 
-    while ((packet_size = read(intf_fd, in_packet, sizeof(in_packet))) > 0) {
+    while ((in_packet_size = read(intf_fd, in_packet, sizeof(in_packet))) > 0) {
         JLG(temp, switch_intf_fd_array, intf_fd);
         hostif_info =
             (switch_hostif_info_t *) (*(unsigned long *)temp);
         if (!hostif_info) {
             perror("invalid hostif fd");
             return;
         }
-//         SWITCH_API_TRACE("Received packet from host interface %lu\n",
-//                          hostif_info->hostif.handle);
-        status = switch_api_hostif_rx_packet_from_host(hostif_info, in_packet,
-                                                    packet_size);
+        SWITCH_API_TRACE("Received packet from host port %s through netdev\n",
+                         hostif_info->hostif.intf_name);
+
+        eth_header = (switch_ethernet_header_t *) in_packet;
+        if (ntohs(eth_header->ether_type) == SWITCH_ETHERTYPE_DOT1Q) {
+            vlan_header = (switch_vlan_header_t *) (in_packet + 2 * ETH_LEN);
+            uint16_t *vlan_h = (uint16_t *) (vlan_header) + 1;
+            vlan_id1 = ntohs(*vlan_h);
+        }
+
+        tx_entry.intf_fd = intf_fd;
+        tx_entry.vlan_id = vlan_id1;
+        status = switch_packet_tx_info_get(
+                             &tx_entry,
+                             &tx_info);
         if (status != SWITCH_STATUS_SUCCESS) {
+            SWITCH_API_ERROR("net filter tx not found. dropping packet");
+            continue;
         }
+
+        memset(&packet_header, 0x0, sizeof(packet_header));
+        cpu_header = &packet_header.cpu_header;
+        fabric_header = &packet_header.fabric_header;
+
+        if (tx_info->bypass_flags == SWITCH_BYPASS_ALL) {
+            cpu_header->tx_bypass = TRUE;
+            cpu_header->reason_code = tx_info->bypass_flags;
+            fabric_header->dst_port_or_group = tx_info->port;
+            memcpy(packet, in_packet, in_packet_size);
+            packet_size = in_packet_size;
+        } else {
+            cpu_header->tx_bypass = FALSE;
+            cpu_header->reason_code = tx_info->bypass_flags;
+            vlan_id1 = tx_info->bd & 0xFFF;
+            vlan_id2 = (tx_info->bd & 0xF000) >> 12;
+
+            eth_header = (switch_ethernet_header_t *) in_packet;
+            ether_type = htons(eth_header->ether_type);
+
+            if (ether_type != SWITCH_ETHERTYPE_DOT1Q) {
+                vlan_offset += sizeof(switch_vlan_header_t);
+            }
+
+            if (vlan_id2) {
```

Comment:
What does vlan_id2 != 0 indicate? This logic is not clear, why vlan_offset is moved based on vlan_id2?